### PR TITLE
Add ability to specify per token expiry in sorted oracles

### LINKED ISSUE #6180: sortedOracles.sol report expiration per token
### Expected Behavior

sortedOracles should have the ability to specify an specific reportExpirySeconds per token.

### Current Behavior

sortedOracles currently only has a global reportExpirySeconds parameter which is used for all tokens.

### PR DESCRIPTION
**Pull request by bowd**
### Description

Currently SortedOracles has a contract wide value for when a report expires. That means that while the contract can be used to support different tokens with different oracles reporting rates, all of the tokens will need to be reported with the same frequency.

We're planning to use SortedOracles as source of truth for the tBTC integration, and we will report Celo/BTC prices there.
During our planning we discussed that having the ability to specify per token report expiry will make the contract more flexible to be used as we were thinking that the BTC oracle will not need to report at the same frequency of the cUSD oracle. 

Thus the changes here are:

- a mapping `address => uint256` where the contract will store the per token expiry rates
- an internal function that looks up a value in the mapping or returns the contact wide default (i.e. what was before the single expiry rate)
- make `isOldestReportExpired` use the function to check for expiry

### Other changes

None

### Tested

Specs update & new spec added to cover the situation when the per token expiry differs from the fallback.

### Backwards compatibility

No braking changes in contract logic, previous expiry becomes the fallback.

### CODE REVIEW

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 1]
```diff
@@ -75,6 +77,21 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
     emit ReportExpirySet(_reportExpirySeconds);
   }
 
+  /**
+   * @notice Sets the report expiry parameter for a token.
+   * @param _token The address of the token to set
+   * @param _reportExpirySeconds The number of seconds before a report is considered expired.
+   */
+  function setTokenReportExpiry(address _token, uint256 _reportExpirySeconds) public onlyOwner {
```
Comment (tkporter) [inline packages/protocol/contracts/stability/SortedOracles.sol]: any reason not external?

**Review by tkporter — APPROVED** [review]
Beautiful

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 2]
```diff
@@ -273,6 +290,19 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
...
+   * @param token The address of the token
+   * @return The report expiry in seconds
+   */
+  function getTokenReportExpirySeconds(address token) public view returns (uint256) {
+    if (tokenReportExpirySeconds[token] == 0) {
+      return reportExpirySeconds;
+    }
+
+    return tokenReportExpirySeconds[token];
+  }
```
Comment (nvtaveras) [inline packages/protocol/contracts/stability/SortedOracles.sol:309]: nit: Do we usually write tests for this kind of functions too?

**Comment by nvtaveras** [comment]
Looks great! 🚢

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 3]
```diff
@@ -273,6 +290,19 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
     return oracles[token];
   }
 
+  /**
+   * @notice Returns the expiry for the token if exists, if not the default
+   * @param token The address of the token
+   * @return The report expiry in seconds
+   */
+  function getTokenReportExpirySeconds(address token) public view returns (uint256) {
+    if (tokenReportExpirySeconds[token] == 0) {
```
Comment (m-chrzan) [inline packages/protocol/contracts/stability/SortedOracles.sol:304]: Is it safe to always fallback on the old default?

File: packages/protocol/test/stability/sortedoracles.ts [hunk 4]
```diff
@@ -179,15 +207,42 @@ contract('SortedOracles', (accounts: string[]) => {
...
+        it('should return false if report is not expired', async () => {
+          const isReportExpired = await sortedOracles.isOldestReportExpired(aToken)
+          assert.isFalse(isReportExpired[0])
+        })
       })
 
-      it('should return false if report is not expired', async () => {
-        const isReportExpired = await sortedOracles.isOldestReportExpired(aToken)
-        assert.isFalse(isReportExpired[0])
+      describe('when a per token expiry is set, which is grater than the default', () => {
```
Comment (m-chrzan) [inline packages/protocol/test/stability/sortedoracles.ts]: Would the mirror case, when the new expiry is lower than the original, be useful?

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 5]
```diff
@@ -75,6 +77,21 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
     emit ReportExpirySet(_reportExpirySeconds);
   }
 
+  /**
+   * @notice Sets the report expiry parameter for a token.
+   * @param _token The address of the token to set
```
Comment (m-chrzan) [inline packages/protocol/contracts/stability/SortedOracles.sol]: Convention nit: period at the end.

Comment (m-chrzan) [reply packages/protocol/contracts/stability/SortedOracles.sol]: Also a wording nit: this sentence feels incomplete, add something like "expiry for" to the end? Or just shorten to "Address of the token".

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 6]
```diff
@@ -273,6 +290,19 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
     return oracles[token];
   }
 
+  /**
+   * @notice Returns the expiry for the token if exists, if not the default
```
Comment (m-chrzan) [inline packages/protocol/contracts/stability/SortedOracles.sol]: Ditto on periods throughout.

File: packages/sdk/contractkit/src/wrappers/SortedOracles.ts [hunk 7]
```diff
@@ -105,6 +105,19 @@ export class SortedOraclesWrapper extends BaseWrapper<SortedOracles> {
     valueToBigNumber
   )
 
+  /**
+   * Returns the token report expiry parameter
```
Comment (m-chrzan) [inline packages/sdk/contractkit/src/wrappers/SortedOracles.ts]: By convention, should this be an exact copy of the comment from the .sol file?

File: packages/protocol/contracts/stability/SortedOracles.sol (hunk 3 above)
Comment (nvtaveras) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: This is a good question. In the way I see it, no token should ever be relying on this value as it feels "unsafe" and not precisely accurate.

I don't know what are the backwards compatibility/upgradeability considerations that we need to have here, but I imagine that every token which the contract is receiving reports for should have a registered report expiry duration.

Comment (bowd) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: The way I see it there's a contract level value which is *always* set because it's part of the constructor, so even if that's `0` it was set intentionally as `0`.

Whereas there's no clear point where one would register a new token that's being handled by an instance of the contract. Therefore there's no way to enforce that a token has a specific expiry set.

So the way I look at it is, we have a global value that we know is always set at contract creation and then we have optional per token values that may or may not be set. This feels safer to me rather than relying on having a mapping for a token created before we start reporting for that token. That's a state that's unenforceable without major changes.

File: packages/protocol/contracts/stability/SortedOracles.sol (hunk 2 above)
Comment (bowd) [reply packages/protocol/contracts/stability/SortedOracles.sol:309]: Yes you're right, I initially had that as internal so I didn't test it, but then decided to expose it to contractkit so I should add tests :+1:

File: packages/protocol/contracts/stability/SortedOracles.sol (hunk 3 above)
Comment (nvtaveras) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: > This feels safer to me rather than relying on having a mapping for a token created before we start reporting for that token. 

What I had in mind is that both the token and the reportExpirySeconds mapping should always be added together in a single governance proposal (and that in the upgrade process for this particular contract we should add the initial mapping for cUSD) which is why to me using a fallback value feels more like a mistake on us forgetting to add the mapping than an intentional side effect.

Comment (bowd) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: I don't really like the fact that we're encoding this dependency: "reportExpirySeconds mapping should always be added together in a single governance proposal" in our heads, but there's no way to encode that explicitly in the contract.

File: packages/protocol/contracts/stability/SortedOracles.sol [hunk 8]
```diff
@@ -28,7 +28,13 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
   mapping(address => mapping(address => bool)) public isOracle;
   mapping(address => address[]) public oracles;
 
+  // `reportExpirySeconds is the fallback value used to determine reporting
```
Comment (nvtaveras) [inline packages/protocol/contracts/stability/SortedOracles.sol]:
```suggestion
  // `reportExpirySeconds` is the fallback value used to determine reporting
```

File: packages/protocol/contracts/stability/SortedOracles.sol (hunk 3 above)
Comment (nvtaveras) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: Had a discussion with @bowd about this. I originally assumed that there was a notion of "adding" a token to the contract before being able to report to it, and hence why I imagined that we would always add a token together with its reportExpirySeconds value. This is not the case, as the only thing needed before being able to report values to a token is to register Oracles to it.

Given that there's no way to enforce that a token must have a reportExpirySeconds I am fine with with keeping things as they are. We also discussed that one could imagine the reportExpirySeconds value being the default for stable tokens native to the protocol (cUSD, cEUR, etc) while the special mapping is only used for special tokens that might have some different requirements.

TL;DR: I'm fine with keeping the current approach unless you have any big concerns @m-chrzan

**Review by nvtaveras — APPROVED** [review]

File: packages/protocol/contracts/stability/SortedOracles.sol (hunk 3 above)
Comment (nambrot) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: Isn't it possible to extend the contract to check when an oracle reports that there is a token-specific expiry?

Comment (m-chrzan) [reply packages/protocol/contracts/stability/SortedOracles.sol:304]: That sounds good to me, thanks for the explanations @bowd and @nvtaveras.
